packages/flutter_tools/lib/src/device.dart

```diff
@@ -915,6 +915,39 @@ class DebuggingOptions {
 
   bool get hasObservatoryPort => hostVmServicePort != null;
 
+  List<String> getIOSLaunchArguments(EnvironmentType environmentType, String? route,  Map<String, Object?> platformArgs) {
+    final String dartVmFlags = computeDartVmFlags(this);
+    return <String>[
+      '--enable-dart-profiling',
+      if (disableServiceAuthCodes) '--disable-service-auth-codes',
+      if (disablePortPublication) '--disable-observatory-publication',
+      if (startPaused) '--start-paused',
+      if (environmentType == EnvironmentType.physical && dartVmFlags.isNotEmpty) '--dart-flags="$dartVmFlags"',
+      if (environmentType == EnvironmentType.simulator && dartVmFlags.isNotEmpty) '--dart-flags=$dartVmFlags',
+      if (useTestFonts) '--use-test-fonts',
+      if (debuggingEnabled) ...<String>[
+        '--enable-checked-mode',
+        '--verify-entry-points',
+      ],
+      if (enableSoftwareRendering) '--enable-software-rendering',
+      if (traceSystrace) '--trace-systrace',
+      if (skiaDeterministicRendering) '--skia-deterministic-rendering',
+      if (traceSkia) '--trace-skia',
+      if (traceAllowlist != null) '--trace-allowlist="$traceAllowlist"',
+      if (traceSkiaAllowlist != null) '--trace-skia-allowlist="$traceSkiaAllowlist"',
+      if (endlessTraceBuffer) '--endless-trace-buffer',
+      if (dumpSkpOnShaderCompilation) '--dump-skp-on-shader-compilation',
+      if (verboseSystemLogs) '--verbose-logging',
+      if (cacheSkSL) '--cache-sksl',
+      if (purgePersistentCache) '--purge-persistent-cache',
+      if (route != null) '--route=$route',
+      if (platformArgs['trace-startup'] as bool? ?? false) '--trace-startup',
+      if (enableImpeller) '--enable-impeller',
+      if (environmentType == EnvironmentType.physical && deviceVmServicePort != null) '--observatory-port=$deviceVmServicePort',
+      if (environmentType == EnvironmentType.simulator && hasObservatoryPort) '--observatory-port=$hostVmServicePort',
```

Comment:
Nit: I know this was my suggestion, but I suspect `hasObservatoryPort` is left over from one of those transitions that needs to be cleaned up (there's a deprecated `flutter run --observatory-port` flag), can you make this null check explicitly for `hostVmServicePort`? This could also use a comment
```suggestion
      // The simulator "device" is actually on the host machine so no ports will be forwarded. Use the suggested host port.
      if (environmentType == EnvironmentType.simulator && hostVmServicePort != null) 
        '--observatory-port=$hostVmServicePort',
```

Reply:
Should I go ahead and remove `hasObservatoryPort`? It's only used on other place.
https://github.com/flutter/flutter/blob/1ca2e0b5b28eb34cb010e8d115f6a4c740de10a0/packages/flutter_tools/lib/src/tester/flutter_tester.dart#L175-L176

Reply:
I think it's wise to remove it, it's not adding that much and is just confusing.